sdk/python/cmd/pulumi-language-python/main.go

```diff
@@ -680,6 +679,7 @@ func (host *pythonLanguageHost) Run(ctx context.Context, req *pulumirpc.RunReque
 			return nil, err
 		}
 	}
+	validateVersion(ctx, host.virtualenvPath)
```

Comment:
Why is this in Run? 

Could we consider placing it right after `prepareVirtualEnvironment` call?

Reply:
All this logic should be done in host start up, not run, really.

Reply:
Eh, I'm not so sure. Did I understand the idea in the PR correctly that validateVersion depends on creating the virtual environment and must run after it?  If this logic goes back to host startup, then we'd need to move prepareVirtualEnvironment to host startup, which may have tricky consequences. Perhaps if we can work out an alternative where validateVersion stays put where it is but gets smarter so it does not fail on empty venv, it would be better. After all creating virtual environments never changes the version of the Python interpreter AFAIK.

Reply:
More for the long term - I am skeptical of logic in host startup based on the experience with java lang provider, where convert started failing because host-startup assumed a project that wasn't there (yet). I've switched it to lazy-instantiation in response to gRPC methods.  If all work happens in response to gRPC methods, there's less coupling of the form "oh this startup logic will always run if I just want to call method XYZ even if XYZ does not need it.." But we might like to have a look at our gRPC methods some time and see if we're missing any, or document better in which sequence they're used.

Reply:
The version check actually doesn't fail. It only prints a warning message which surprised me when looking into this.

Reply:
SO the question is could you alter that and still get a satisfactory solution without moving the code to Run?

Reply:
> More for the long term - I am skeptical of logic in host startup based on the experience with java lang provider, 

That's fair, I was thinking of "shimless" where we might not be calling Run but still need the venv setup. I guess it's just a case of making sure all the different entrypoints correctly call setup logic before actually running.

Reply:
the motivation for putting the python version check here was because I wanted to make sure that anything that might happen around configuring a pulumi python program was settled before doing the version warning.

Putting it in GetRequiredPlugins means that we depend on running GetRequiredPlugins in order to warn the user that their version of python may not work.

I've moved the code according to your suggestion, but that was the motivation around my first choice.